### PR TITLE
ARROW-14899: [C++] implement GcsInputStream::GetMetadata

### PR DESCRIPTION
**Pull request by coryan**

### CODE REVIEW

**Review by scotthart — APPROVED** [review]

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 1]
```diff
@@ -352,6 +360,68 @@ TEST_F(GcsIntegrationTest, ReadObjectInfoInvalid) {
   EXPECT_EQ(result.status().code(), StatusCode::IOError);
 }
```
Comment (pitrou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:362]: Is it also possible to add a simple roundtrip test between `ToObjectMetadata` and `FromObjectMetadata`?

**Review by pitrou — APPROVED** [review]
LGTM, just a suggestion

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 1 above)
Comment (coryan) [reply cpp/src/arrow/filesystem/gcsfs_test.cc:391]: Done (with some limitations, not all fields have public modifiers in `gcs::ObjectMetadata`)

**Review by pitrou — APPROVED** [review]
Thanks for the update!

**Pull request closed without merging**